### PR TITLE
Show the Contao layout in the Symfony profiler

### PR DESCRIPTION
**Pull request by aschempp**
Bildschirmfoto 2021-04-07 um 11 19 27

### CODE REVIEW

**Review by fritzmg — DISMISSED** [review]

File: core-bundle/src/Resources/contao/elements/ContentModule.php [hunk 1]
```diff
@@ -62,6 +62,12 @@ public function generate()
 		$objModel->cssID = $cssID;
 		$objModel->typePrefix = 'ce_';
 
+		if (System::getContainer()->getParameter('kernel.debug'))
+		{
+			$objStopwatch = System::getContainer()->get('debug.stopwatch');
```
Comment (ausi) [inline core-bundle/src/Resources/contao/elements/ContentModule.php:85]: Is this service guaranteed to be always there if `kernel.debug` is `true`? Even in custom Symfony applications where Contao is only used as a bundle?

Comment (aschempp) [reply core-bundle/src/Resources/contao/elements/ContentModule.php:85]: It's always there if the `Stopwatch` class exists (https://github.com/symfony/symfony/blob/5.x/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L938) and stopwatch is required by Contao (https://github.com/contao/contao/blob/4.9/composer.json#L107)

**Comment by Toflar** [comment]
a) Looks like a feature, no?
b) Does that work with our new fragments?

**Comment by aschempp** [comment]
> a) Looks like a feature, no?

Kind of, yes. But it's a debug improvement which would be useful in our LTS.

> b) Does that work with our new fragments?

I think it does. Since we profile in `getFrontendModule`, whatever that method does (e.g. generate a fragment) will be included in the profile. Even if the fragment would be ESI-rendered, be aware that the debug mode does not have a cache, therefore no ESI support.

**Review by leofeyer — CHANGES_REQUESTED** [review]
The stopwatch is currently started before the module/content class is generated, which means constructing the object is part of the count. Is this correct or should we rather only count the time of the `->generate()` method?

**Comment by bytehead** [comment]
> only count the time of the `->generate()` method

👍

**Comment by aschempp** [comment]
I would count everything, because everything is related to that element/module/controller. If the constructor doesn't do much, it doesn't matter. If the constructor is huge, I would want it to appear in that profile.

File: core-bundle/src/Resources/contao/elements/ContentModule.php [hunk 2]
```diff
@@ -72,7 +78,14 @@ public function generate()
 			$responseTagger->addTags(array('contao.db.tl_content.' . $this->id));
 		}
 
-		return $objModule->generate();
+		$strBuffer = $objModule->generate();
+
+		if ($objStopwatch)
```
[the remark this answers is not in the document]
Comment (aschempp) [reply core-bundle/src/Resources/contao/elements/ContentModule.php]: Should I use `isset()` or `instanceof` instead?

[the remark this answers is not in the document]
Comment (leofeyer) [reply core-bundle/src/Resources/contao/elements/ContentModule.php]: I think you just have to add `$objStopwatch = null;` above so the variable is always defined.

**Comment by aschempp** [comment]
What's necessary to get this merged?

**Comment by leofeyer** [comment]
Fixing the failing CI jobs and addressing the change requests. 😉

File: core-bundle/src/Resources/contao/elements/ContentModule.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (fritzmg) [reply core-bundle/src/Resources/contao/elements/ContentModule.php]: I would prefer `if (isset($objStopwatch))` in order to have less additional lines of code.

**Review by ausi — DISMISSED** [review]

File: core-bundle/src/Resources/contao/elements/ContentModule.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (leofeyer) [reply core-bundle/src/Resources/contao/elements/ContentModule.php]: Fixed in a22fa63fee3d8c139a3c04669804833fddff37c2.

File: core-bundle/src/Resources/contao/elements/ContentModule.php (hunk 1 above)
Comment (fritzmg) [reply core-bundle/src/Resources/contao/elements/ContentModule.php:85]: ~Curiously I have a Contao Manged Edition `4.9.16` installation where I now get this error:~

// nvm. the instance had an erroneous `Controller.php`

**Comment by bytehead** [comment]
Updated to 4.9.17 yesterday evening and I get this error:
`Event "contao.frontend_module.header (ID 10)" is not started.` maybe related?

**Pull request merged**